### DISCUSSION TITLE
Running Chigger to extract images from output files

### DISCUSSION THREAD

**Discussion by Amir1361** [idaholab/moose#17320] [Q&A Modules: Phase field] [answered]
Hi

I wrote a python script to extract the images from MOOSE output files. The code works well on my personal computer (Linux), but I get a Segmentation fault when I run it on a cluster computer. I use Chigger as below:

reader = chigger.exodus.ExodusReader(filepath)
reader.setOptions(block=['0'])

result = chigger.exodus.ExodusResult(reader)
result.setOptions(edge_color=[0, 0, 0], variable='c1', block=['0'], cmap='grayscale', local_range=True, camera=camera)

window = chigger.RenderWindow(result, offscreen=True)
print("after chigger.RenderWindow")
imagename =”%s_100hr.jpg"%folder[4:]
#window.start()
print("before window.write, imagename is ",imagename)
window.write(imagename)
print("after chigger.RenderWindow")

The segmentation error comes up with "window.write(imagename)". Could you please help me to solve this problem?
Thanks
Amir

**Comment by aeslaughter** [comment, marked as answer]
chigger needs an active display for rendering. In theory there should be a way to compile VTK to operate using software rendering. We have tried several times over the years to get that working, but have failed in each attempt.

**Reply by Amir1361** [reply]
Thank you for your reply. If I understood you correctly, I need to have GUI to run Chigger. So, using offscreen=true in window = chigger.RenderWindow(result, offscreen=True) is not enough, and we will get a segmentation error.

**Reply by aeslaughter** [reply]
Correct. The "offscreen=True" just doesn't show the window, but still requires openGL and a screen. I need to improve the error message for this case. Better yet figure out how to make it work, but that will be awhile, if ever.

**Reply by Amir1361** [reply]
Got it! Thank you very much!

**Comment by Amir1361** [comment]
@aeslaughter 

Hi Andrew,

I am running the python code based on chigger to extract microstructures from MOOSE output files. I have more than 2000 output folders with 61 output files inside them. In fact, I want to extract more than 120,000 images. The code is doing well on the small number of output directories (less than 50). But when I want to run it for all 2000 directories, I get the following error message:

    with open(filename, "a+") as f: # "a+" to make sure it gets created
    OSError: [Errno 5] Input/output error: '/run/user/1001/gvfs/smb-share:server=cifs-prd-01,share=research/MahmoodMamivand/common/CMDLab/Fe_Cr_Co/time_temp_comp/MOOSE_results/run_cr0.05_co0.35_T870/FeCrCo_cr0.05_co0.35_T870_exodus.e-s043'

    Exodus Library Warning/Error: [vtkexodusII_ex__check_valid_file_id]
	ERROR: In "vtkexodusII_ex_close", the file id 65536 was not obtained via a call to "ex_open" or "ex_create".
		It does not refer to a valid open exodus file.
		Aborting to avoid file corruption or data loss or other potential problems.

65536 is the maximum 16-bit integer + 1, so I'm hitting a limit somewhere. Does the chigger interface have a method to clean up its internals? What is your suggestion? Thank you very much for your consideration,

Sincerely,
Amir
[output.txt](https://github.com/idaholab/moose/files/6835231/output.txt)

**Reply by aeslaughter** [reply]
This is some sort of system limitation and an unsure of the proper fix, I am wondering if it is as easy as changing the `ulimit`. @milljm has more experience in this realm, perhaps he can chime in.

**Reply by Amir1361** [reply]
Thanks for your response. I tried  `ulimit` before. It doesn't work for me.

**Reply by milljm** [reply]
You might want to try offloading the data onto a hard disk location (if I am reading that mount point correctly /media). Is that a thumb stick?

Or, try moving the data to a Unix filesystem such as EXT (if it's in fact a SMB mount to a Windows share). The last I heard (which was a long time ago, and I am probably wrong), is reading/writing to NTFS filesystems and the like was experimental.

**Reply by Amir1361** [reply]
Thank you Jason, I read the output files from Windows share storage (SMB). It's interesting because it works well when I run it on a small number of output directories (less than 50). I will test your suggestions. Thanks again!
